scipy/optimize/minpack.py

```diff
@@ -606,13 +608,16 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         args, varargs, varkw, defaults = _getargspec(f)
         if len(args) < 2:
             raise ValueError("Unable to determine number of fit parameters.")
-        p0 = np.ones(len(args) - 1)
+        n = len(args) - 1
     else:
         p0 = np.atleast_1d(p0)
+        n = p0.size
 
-    lb, ub = prepare_bounds(bounds, p0)
-    bounded_problem = np.any((lb > -np.inf) | (ub < np.inf))
+    lb, ub = prepare_bounds(bounds, n)
+    if p0 is None:
+        p0 = _initialize_feasible(lb, ub)
```

Comment:
So the initial point of `curve_fit(f, x, y)`, which uses LM, is `[1]*n` because `_initalize_feasible` returns all ones for unbounded problems. And if at some point `_initialize_feasible` is refactored, the starting point for fitting without bounds suddenly has a different starting point... 

OTOH, one may argue that users should not rely on the implementation detail of what the default starting point is. So it's probably fine